python_modules/dagster/dagster_tests/core_tests/test_utils.py

```diff
@@ -142,3 +142,11 @@ def sleepy_thread():
 
         assert executor.num_running_futures == 0
         assert executor.num_queued_futures == 0
+
+        # futures still have strong refs so are still tracked
+        assert executor.weak_tracked_futures_count == 10
+
+        futures = []
+        f = None
```

Comment:
Looks like we kept a strong reference to a remaining single future when iterating on a list of futures — this potentially means that in the downstream executor implementation, when do the same thing, we’ll be off by one if we don’t explicitly remove the reference to the future?

Reply:
> this potentially means that in the downstream executor implementation, when do the same thing, we’ll be off by one if we don’t explicitly remove the reference to the future?

While the local variable is in still in scope (like in this case since we are still in the same scope as `for f in as_completed(futures)` a strong ref still exists, but I do not think this presents a systemic issue. It should get still get cleaned up when the function its used in exits or the next iteration if its inside of a long lived loop. 

The two concerns we have here are 
* how many items we have to iterate over to calculate `num_running_futures` 
* how much we accumulate in memory over time

Cases like "for loop var exists for entire scope" will cause there to one extra item which wont meaningfully impact either concern.